src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -0,0 +1,337 @@
+        }
+
+        private void AddDiagnostics(ImmutableArray<Diagnostic> diagnostics)
+        {
+            // Atomic update operation. Applies a function (Concat) to a variable repeatedly, until it "gets through" (isn't in a race condition with another concat)
+            var oldDiags = _diagnostics;
+            while (true)
+            {
+                var newDiags = oldDiags.IsDefault ? diagnostics : oldDiags.Concat(diagnostics);
+                var overwriteDiags = ImmutableInterlocked.InterlockedCompareExchange(ref _diagnostics, newDiags, oldDiags);
+                if (overwriteDiags == oldDiags)
+                {
+                    break;
+                }
+                oldDiags = overwriteDiags;
+            }
+        }
+
+        public override bool IsVararg
+        {
+            get
+            {
+                ComputeParameters();
+                return _isVararg;
+            }
+        }
+
+        public override ImmutableArray<ParameterSymbol> Parameters
+        {
+            get
+            {
+                ComputeParameters();
+                return _parameters;
+            }
+        }
+
+        private void ComputeParameters()
+        {
+            if (!_parameters.IsDefault)
+            {
+                return;
+            }
+            var diagnostics = DiagnosticBag.GetInstance();
+            SyntaxToken arglistToken;
+            _parameters = ParameterHelpers.MakeParameters(_binder, this, _syntax.ParameterList, true, out arglistToken, diagnostics);
+            _isVararg = (arglistToken.Kind() == SyntaxKind.ArgListKeyword);
+            AddDiagnostics(diagnostics.ToReadOnlyAndFree());
+        }
+
+        public override TypeSymbol ReturnType
+        {
+            get
+            {
+                ComputeReturnType();
+                return _returnType;
+            }
+        }
+
+        private void ComputeReturnType()
+        {
+            if (_returnType != null)
+            {
+                return;
+            }
+            var diagnostics = DiagnosticBag.GetInstance();
+            bool isVar;
+            _returnType = _binder.BindType(_syntax.ReturnType, diagnostics, out isVar);
+            if (isVar)
+            {
+                // TODO: Add a better message for this
+                diagnostics.Add(ErrorCode.ERR_RecursivelyTypedVariable, _syntax.ReturnType.Location, this);
+                _returnType = _binder.CreateErrorType("var");
+            }
+            AddDiagnostics(diagnostics.ToReadOnlyAndFree());
+        }
+
+        public override bool ReturnsVoid => ReturnType.SpecialType == SpecialType.System_Void;
+
+        public override int Arity => TypeParameters.Length;
+
+        public override ImmutableArray<TypeSymbol> TypeArguments => TypeParameters.Cast<TypeParameterSymbol, TypeSymbol>();
+
+        public override ImmutableArray<TypeParameterSymbol> TypeParameters => _typeParameters;
+
+        public override bool IsExtensionMethod
```

Comment:
We do not intend to support local extension methods (they can't be static, for example)